tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/TableauEncoder.scala

```diff
@@ -0,0 +1,424 @@
+    extends LazyLogging {
+  val levelFinder = new TlaLevelFinder(module)
+  val varNamesToExStrings = new HashMap[String, String]()
+
+  /**
+   * Encodes each of a sequence of temporal formulas.
+   * @see
+   *   [[encodeFormula]]
+   */
+  def temporalsToInvariants(
+      modWithPreds: ModWithPreds,
+      formulas: Seq[TlaOperDecl]): ModWithPreds = {
+
+    formulas.foldLeft(modWithPreds)(singleTemporalToInvariant)
+  }
+
+  /**
+   * Adds a variable that stores the initial value of the given expression ex. Create a new variable Init_ex and modify
+   * init and next:
+   *
+   * newInit == oldInit /\ Init_ex = ex
+   *
+   * newNext == oldNext /\ UNCHANGED << Init_ex >>
+   */
+  def addInitExVar(modWithPreds: ModWithPreds, ex: TlaEx, exName: String): (ModWithPreds, TlaVarDecl) = {
+    val exVarDecl = new TlaVarDecl("__" + exName + "_init")(Typed(BoolT1))
+    val exVar = builder.declAsNameEx(exVarDecl)
+
+    val newInit = andInDecl(
+        builder.eql(
+            exVar,
+            builder.useTrustedEx(ex),
+        ),
+        modWithPreds.init,
+        tracker,
+    )
+
+    val newNext = andInDecl(
+        builder.unchanged(exVar),
+        modWithPreds.next,
+        tracker,
+    )
+
+    val curModWithPreds = modWithPreds.setPredicates(newInit, newNext, modWithPreds.loopOK)
+    (curModWithPreds.prependDecl(exVarDecl), exVarDecl)
+  }
+
+  /**
+   * Encodes a given formula, using the Tableau encoding by adjusting init, next, loopOK and the given formula.
+   */
+  def singleTemporalToInvariant(
+      modWithPreds: ModWithPreds,
+      formula: TlaOperDecl): ModWithPreds = {
+
+    var (curModWithPreds, formulaEx) = encodeSyntaxTreeInPredicates(modWithPreds, formula.body)
+
+    val tuple = addInitExVar(curModWithPreds, formulaEx, formula.name)
+    curModWithPreds = tuple._1
+    val initExVarDecl = tuple._2
+
+    // replace formula by formula == loopOK => formula_init, and move formula to the end of the spec so it is after loopOK
+    val newFormulaDecl = new TlaOperDecl(
+        formula.name,
+        formula.formalParams,
+        builder.impl(
+            builder.appOp(builder.declAsNameEx(curModWithPreds.loopOK), Seq.empty[TBuilderInstruction]: _*),
+            builder.declAsNameEx(initExVarDecl),
+        ),
+    )(formula.typeTag)
+    val newDecls = curModWithPreds.module.declarations.filterNot(decl => decl.name == formula.name) :+ newFormulaDecl
+    val newModule = new TlaModule(curModWithPreds.module.name, newDecls)
+
+    curModWithPreds.setModule(newModule)
+  }
+
+  /**
+   * Takes a temporal operator and a name, and generates a new variable declaration declaring the auxiliary "unrolling"
+   * variable for that temporal operator.
+   */
+  private def createUnrollingVar(oper: TlaTempOper, nodeIdentifier: String): TlaVarDecl = {
+    val nameSuffix = oper match {
+      case TlaTempOper.box     => TableauEncoder.BOX_SUFFIX
+      case TlaTempOper.diamond => TableauEncoder.DIAMOND_SUFFIX
+    }
```

Comment:
AFAICT, this should produce a warning due to being underspecified for other `TlaTempOper`. You might need to add a default case with a comment to the effect of "impossible by construction".